doc/reference/trd-hil-design.md

```diff
@@ -0,0 +1,685 @@
+you are working with a software one, you are precluding hardware
+implementations from using your HIL. This means that a developer must
+decide to use either the blocking HIL (which in some cases can't exist)
+or the non-blocking one, making software less reusable.
+
+Rule 1: Don't Make Synchronous Callbacks
+===============================
+
+Consider the following API for requesting 32 bits of randomness:
+
+```rust
+trait Random {
+  fn random(&self) -> Result<(), ErrorCode>;
+  fn set_client(&self, client: &'static Client);
+}
+
+trait Client {
+  fn random_ready(&self, result: Result<u32, ErrorCode>);
+}
+```
+
+If `Random` is implemented on top of a hardware random number
+generator, the random bits might not be ready until an interrupt
+is issued. E.g., if the implementation generates random numbers
+by running AES128 in counter mode on a hidden seed[HCG], then 
+generating random bits may require an interrupt.
+
+But AES128 computes *4* 32-bit values of randomness. So a smart
+implementation will compute 128 bits, and call back with 32 of them.
+The next 3 calls to `random` can produce data from the remaining
+data. The simple implementation for this algorithm is to call `random_ready` 
+inside the call to `random` if cached results are ready: the values 
+are ready, so issue the callback immediately.
+
+Making the `random_ready` callback from inside `random` is a bad idea for
+two reasons: call loops and client code complexity.
+
+The first issue that arises is it can create call loops. Suppose that 
+the client wants 1024 bits (so 32 words) or randomness. It needs to
+invoke `random` 32 times. The standard call pattern is to call `random`,
+then in the `random_ready` callback, store the new random bits and call
+`random` again. This repeats 32 times. 
+
+If the implementation uses an interrupt every 4 calls, then this call 
+pattern isn't terrible: it would result in 8 stack frames. But suppose
+that the implementation chooses to generate not 128 bits at a time, but
+rather 1024 bits (e.g., runs counter mode on 32 words). Then one could have
+up to 64 stack frames. It might be that the compiler inlines this, but it
+also might not. Assuming the compiler always does a specific optimization 
+for you is dangerous: there all sorts of edge cases and heuristics, and
+trying to adjust source code to coax it to do what you want (which can
+change with each compiler release) is brittle.
+
+The second, and more dangerously, client logic becomes much more complex.
+For example, consider this client code:
+
+```rust
+  let result = random.random();
+  match result {
+    Ok(()) => self.state.set(State::Waiting),
+	Err(e) => self.state.set(State::Error),
+  }
+  
+fn random_ready(&self, bits: u32, result: Result<(), ErrorCode>) {
+  match result {
+     Ok(()) => {
+	   // Use the random bits
+	   self.state.set(State::Idle);
+     },
+	 Err(e) => {
+	   self.state.set(State::Error);
+	 }
```

Comment:
Something is off with this code block's formatting. It seems to mix indentations widths of 2, 5, and 8.

Reply:
Yes, I think we should format & squash prior to merge, as there's trailing whitespaces in the doc and the 80 character limit is commonly violated when using the GitHub suggestions.

Reply:
I think I have fixed the code blocks, I will check them.

Reply:
The blocks were mostly fixed, I've run them all through rustfmt to check and applied the very few remaining fixes.